kv/txn_coord_sender.go

```diff
@@ -802,3 +783,61 @@ func (tc *TxnCoordSender) heartbeat(id string) {
 		}
 	}
 }
+
+func (tc *TxnCoordSender) heartbeat(id string, trace *tracer.Trace, ctx context.Context) bool {
+	tc.Lock()
+	proceed := true
+	txnMeta := tc.txns[id]
+	// Before we send a heartbeat, determine whether this transaction
+	// should be considered abandoned. If so, exit heartbeat.
+	if txnMeta.hasClientAbandonedCoord(tc.clock.PhysicalNow()) {
+		// TODO(tschottdorf): should we be more proactive here?
+		// The client might be continuing the transaction
+		// through another coordinator, but in the most likely
+		// case it's just gone and the open transaction record
+		// could block concurrent operations.
+		if log.V(1) {
+			log.Infof("transaction %s abandoned; stopping heartbeat",
+				txnMeta.txn)
+		}
+		proceed = false
+	}
+	// txnMeta.txn is possibly replaced concurrently,
+	// so grab a copy before unlocking.
+	txn := txnMeta.txn
+	tc.Unlock()
+	if !proceed {
+		tc.unregisterTxn(id)
+		return false
+	}
+
+	request := &proto.HeartbeatTxnRequest{
+		RequestHeader: proto.RequestHeader{
+			Key:  txn.Key,
+			User: security.RootUser,
+			Txn:  &txn,
+		},
+	}
+
+	request.Header().Timestamp = tc.clock.Now()
+	reply := &proto.HeartbeatTxnResponse{}
+	call := proto.Call{
+		Args:  request,
+		Reply: reply,
+	}
+
+	epochEnds := trace.Epoch("heartbeat")
+	tc.wrapped.Send(ctx, call)
+	epochEnds()
+	// If the transaction is not in pending state, then we can stop
+	// the heartbeat. It's either aborted or committed, and we resolve
+	// write intents accordingly.
+	if reply.GoError() != nil {
+		log.Warningf("heartbeat to %s failed: %s", txn, reply.GoError())
+	} else if reply.Txn != nil && reply.Txn.Status != proto.PENDING {
+		// The transaction has been committed or aborted, so no use
```

Comment:
Right. I put the `unregisterTxn` call back where it was, and removed the new test I added (which was getting to be tricky to synchronize with the black-box heartbeat loop).